L1Trigger/Phase2L1ParticleFlow/src/BJetId.cc

```diff
@@ -0,0 +1,95 @@
+}
+
+BJetId::~BJetId() { tensorflow::closeSession(session_); }
+void BJetId::setNNVectorVar() {
+  NNvectorVar_.clear();
+  for (int i0 = 0; i0 < fNParticles_; i0++) {
+    if (fPt_.get()[i0] == 0) {
+      for (int i1 = 0; i1 < 13; i1++)
+        NNvectorVar_.push_back(0);
+      continue;
+    }
+    NNvectorVar_.push_back(fId_.get()[i0] == l1t::PFCandidate::Electron && fCharge_.get()[i0]<0);       // Electron
+    NNvectorVar_.push_back(fId_.get()[i0] == l1t::PFCandidate::Electron && fCharge_.get()[i0]>0);       // Positron
+    NNvectorVar_.push_back(fId_.get()[i0] == l1t::PFCandidate::Muon && fCharge_.get()[i0]<0);           // Muon
+    NNvectorVar_.push_back(fId_.get()[i0] == l1t::PFCandidate::Muon && fCharge_.get()[i0]>0);           // Anti-Muon
+    NNvectorVar_.push_back(fId_.get()[i0] == l1t::PFCandidate::Photon && fCharge_.get()[i0]==0);         // Photon
+    NNvectorVar_.push_back(fId_.get()[i0] == l1t::PFCandidate::NeutralHadron && fCharge_.get()[i0]==0);  // Neutral Had
+    NNvectorVar_.push_back(fId_.get()[i0] == l1t::PFCandidate::ChargedHadron && fCharge_.get()[i0]<0);  // Pion
+    NNvectorVar_.push_back(fId_.get()[i0] == l1t::PFCandidate::ChargedHadron && fCharge_.get()[i0]>0);  // Anti-Pion
+    NNvectorVar_.push_back(fDZ_.get()[i0]);   //dZ
+    NNvectorVar_.push_back(std::sqrt(fDX_.get()[i0]*fDX_.get()[i0] + fDY_.get()[i0]*fDY_.get()[i0]));  //d0
+    NNvectorVar_.push_back(fPt_.get()[i0]);   //pT as a fraction of jet pT
+    NNvectorVar_.push_back(fEta_.get()[i0]);  //dEta from jet axis
+    NNvectorVar_.push_back(fPhi_.get()[i0]);  //dPhi from jet axis
+  }
+}
+float BJetId::EvaluateNN() {
+  tensorflow::Tensor input(tensorflow::DT_FLOAT,
+                           {1, (unsigned int)NNvectorVar_.size(), 1});
+  for (unsigned int i = 0; i < NNvectorVar_.size(); i++) {
+    input.tensor<float,3>()(0, i, 0) = float(NNvectorVar_[i]);
+  }
+  std::vector<tensorflow::Tensor> outputs;
+  tensorflow::run(session_, {{fInput_, input}}, {"sequential/dense_2/Sigmoid"}, &outputs);
+  //tensorflow::run(session_, {{fInput_, input}}, {"sequential/sigmoid/Sigmoid"}, &outputs);//TODO: for QKeras models, add auto switch
+  return outputs[0].matrix<float>()(0,0);
+}  //end EvaluateNN
+
+float BJetId::compute(const l1t::PFJet &iJet, float vz, bool useRawPt) {
+  for (int i0 = 0; i0 < fNParticles_; i0++) {
+    fPt_.get()[i0] = 0;
+    fEta_.get()[i0] = 0;
+    fPhi_.get()[i0] = 0;
+    fId_.get()[i0] = 0;
+    fCharge_.get()[i0] = 0;
+    fDZ_.get()[i0] = 0;
+    fDX_.get()[i0] = 0;
+    fDY_.get()[i0] = 0;
+  }
+  auto iParts = iJet.constituents();
+  std::sort(iParts.begin(), iParts.end(), [](edm::Ptr<l1t::PFCandidate> i, edm::Ptr<l1t::PFCandidate> j) { return (i->pt() > j->pt()); });
+  float jetpt;
+  if (useRawPt) {
+    jetpt = iJet.rawPt();
+  } else {
+    jetpt = iJet.pt();
+  }
+  for (unsigned int i0 = 0; i0 < iParts.size(); i0++) {
+    if (i0 > n_particles_max || i0 >= (unsigned int)fNParticles_)
+      break;
```

Comment:
I think here it would be cleaner do to
`for (unsigned int i0 = 0, n = std::min<unsigned int>(iParts.size(), fNParticles_); i0 < n0; ++i0) {`
and have the constructor make sure `fNParticles_` doesn't exceed `n_particles_max`